---
title: "Open Tree of Life Trees"
output:
  html_document:
    df_print: paged
---

https://opentreeoflife.github.io/SSBworkshop/


## Load packages

```{r}
library(rotl)
library(datelife)# devtools::install_github("phylotastic/datelife")
library(dplyr)
library(ggplot2)
library(ggtree)
library(ggstance)#horizontal plots
library("tidytree")
library(viridis)
```


## Define lineages ofinterest

```{r}
my_taxa <- c("Arthropoda")#Match taxonomic names to the Open Tree Taxonomy.

resolved_names <- rotl::tnrs_match_names(names = my_taxa)
resolved_names
```



## Arthropoda Tree 

- At order level

```{r message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
Arthropoda_children <- datelife::get_ott_children(ott_ids = resolved_names[resolved_names$unique_name=="Arthropoda",]$ott_id, ott_rank = "order")

Arthropoda_species <-Arthropoda_children$Arthropoda[Arthropoda_children$Arthropoda$rank=="order",] # protura has manly spp bu tno orders

#Crustacea_species_subtree <- rotl::tol_induced_subtree(Crustacea_species$Arthropoda$ott_id)
Arthropoda_species_subtree <- rotl::tol_induced_subtree(Arthropoda_species$ott_id)
```




```{r message=FALSE, warning=FALSE}
Arthropoda_species_subtree
```



```{r, fig.height=16, fig.width=12}
#ape::plot.phylo(Crustacea_species_subtree, show.node.label	 = T, cex = 1.2)

GgTree_Arthropoda_species<-ggtree(Arthropoda_species_subtree) +ggplot2::xlim(0, 24)+
  geom_text(aes(label=label), size=3, color="Blue", hjust=-0.01)

GgTree_Arthropoda_species
```


<!-- ### Save tree -->

<!-- ```{r} -->
<!--  ape::write.tree(Crustacea_species_subtree, file='Crustacea_species_subtree.txt')  -->
<!-- ``` -->

## Load genomes info

```{r message=FALSE, warning=FALSE}
Arthropods_genomes<-readr::read_csv("Arthropoda_genome_info_withTAXONOMY_06.06.csv") 
```

- Select MOST RECENT assembly per spp


```{r}
Arthropoda_genomes_firstperspp<-Arthropods_genomes %>% 
  filter(phylum=="Arthropoda") %>% 
  arrange(spp, sub.date ) %>% 
  group_by(spp) %>% 
  slice(1)
```

- Are all orders in my tree?

```{r}
Uniqueorders_genomes<-Arthropods_genomes %>% 
  select(order) %>% 
  distinct()
```

```{r}
table_info <- data.frame(label = Arthropoda_species_subtree$tip.label, 
                order_ot_0= sapply(stringr::str_split(Arthropoda_species_subtree$tip.label, "_ott",  n = 2), `[`, 1))

table_info$order_ot<-sapply(stringr::str_split(table_info$order_ot_0, "_"), `[`, 1)

#table_info$spp<-stringr::str_replace(table_info$order_ot_0, "_")

  
table_info
```




- Orders in genome table present in the tree?

```{r}
table(Uniqueorders_genomes$order%in%table_info$order_ot)


Uniqueorders_genomes$order[!Uniqueorders_genomes$order%in%table_info$order_ot]
```


- Orders in the tree present in the genome table

```{r}
table(table_info$order_ot%in%Uniqueorders_genomes$order)
```
```{r, fig.height=20, fig.width=18}
ggtree_Arthropoda_species_subtree_all_labels<-ggtree(Arthropoda_species_subtree) %<+% table_info +ggplot2::xlim(0, 24)+
  geom_text(aes(label=label), size=3, color="black", fontface="bold", hjust=-0.01)+ theme_tree2(legend.position='right') 


ggtree_Arthropoda_species_subtree_all_labels

ggsave(ggtree_Arthropoda_species_subtree_all_labels, file="figs/ggtree_Arthropoda_species_subtree_all_labels.svg", height=18, width=16, units="in")
```

Deleting bad tip

```{r}
Arthropoda_species_subtree <- treeio::drop.tip(Arthropoda_species_subtree, "mrcaott47716ott167281" )
```


```{r, fig.height=20, fig.width=18}
ggtree_Arthropoda_species_subtree<-ggtree(Arthropoda_species_subtree) %<+% table_info +ggplot2::xlim(0, 24)+
        geom_tiplab(aes(label=order_ot), parse=T, size=3, color="Blue", hjust=-0.01)

ggtree_Arthropoda_species_subtree

ggsave(ggtree_Arthropoda_species_subtree, file="figs/ggtree_Arthropoda_species_subtree.svg", height=18, width=16, units="in")
```
 

- One strange tip "mrcaott47716ott167281". Corresponds to [Septosaccus rodriguezii](https://tree.opentreeoflife.org/opentree/argus/opentree12.3@ott167281) are broken taxa.. 

## Add genome data to tree


### Genome lengths per order


```{r, fig.height=18, fig.width=9}


table_info_genomelen<-table_info %>% left_join(Arthropoda_genomes_firstperspp , by=c("order_ot"="order"), keep=TRUE)


Glen_order_facet<-ggtree_Arthropoda_species_subtree+ xlim_tree(25)+
                  geom_facet( data=table_info_genomelen, panel="Genome Len",
                      geom = geom_jitter ,     #geom_point2
                      mapping = aes(x = seq_length/1000000000), color = "blue",   stat='identity',position= position_jitter(height=0.5,width=0) )+
                      scale_color_brewer(palette = "Set1")+
                  geom_vline( data = table_info_genomelen, aes(xintercept=1), colour="red") +labs(x="Genome Assembly Length (Gb)")
Glen_order_facet  
```



### Number of genomes per order


```{r}
Numgenomes_per_order<-Arthropoda_genomes_firstperspp %>% 
              group_by(order) %>% 
              filter(is.na(GENOME_GFF_size)) %>%
              mutate(NumberGenomes=n() ) %>% 
              select(order, NumberGenomes ) %>% 
              distinct(order, .keep_all=TRUE)

Numgenomes_annotated_per_order<-Arthropoda_genomes_firstperspp %>% 
              group_by(order) %>% 
              filter(GENOME_GFF_size > 1) %>%
              mutate(NumberGenomesAnnotated=n() ) %>% 
              select(order, NumberGenomesAnnotated ) %>% 
              distinct(order, .keep_all=TRUE)

table_info_numgenomes<-table_info %>%
  left_join(Numgenomes_per_order , by=c("order_ot"="order"), keep=TRUE) %>%
  left_join(Numgenomes_annotated_per_order, by=c("order_ot"="order"), keep=TRUE) %>%
  mutate(NumberGenomes=tidyr::replace_na(NumberGenomes, replace=0)) %>% 
  mutate(NumberGenomesAnnotated=tidyr::replace_na(NumberGenomesAnnotated, replace=0)) %>%
  mutate(NumberGenomes_2=NumberGenomes) %>%
  mutate(NumberGenomesAnnotated_2=NumberGenomesAnnotated)


table_info_numgenomes_fortree<-table_info_numgenomes %>% select(-NumberGenomes) %>% select(-NumberGenomesAnnotated)  ## can't be same name in tree and in facet or crashes
table_info_numgenomes_forfacet<-table_info_numgenomes %>% select(-NumberGenomes_2) %>% select(-NumberGenomesAnnotated_2)  ## can't be same name in tree and in facet or crashes

```

## Orders - Num genomes - Genome lenght

- Adding the genome lengths into the tree to print them as text

```{r}

Arthropoda_species_Annot<-treeio::full_join(as_tibble(Arthropoda_species_subtree), as_tibble(table_info_numgenomes_fortree), by=c("label"="label"))
Arthropoda_species_Annot<-Arthropoda_species_Annot[!grepl("mrcaott47716ott167281",Arthropoda_species_Annot$label),]
```

## Pivot data from wide to long

```{r}
table_info_numgenomes_forfacet <- table_info_numgenomes_forfacet %>% tidyr::pivot_longer(cols=c("NumberGenomesAnnotated","NumberGenomes"), values_to="Values", names_to = "genomes_type",)

```

```{r, fig.height=18, fig.width=12}


ggtree_Arthropoda_species_3panels<- ggtree(as.treedata(Arthropoda_species_Annot))+ xlim_tree(70)+
  geom_text(aes(label=order_ot), size=4.5, color="black",family="Arial", hjust=-0.01, )+
  geom_tiplab(aes(label = NumberGenomes_2), color = "black", size = 4, offset = 50) +
  geom_tiplab(aes(label = NumberGenomesAnnotated_2), color = "black", size = 4, offset = 40) +
  geom_facet(data=table_info_numgenomes_forfacet, panel="Number of genomes",
                        geom = geom_bar,     #geom_point2
                        mapping = aes(x = Values, fill = genomes_type), show.legend=TRUE,  stat='identity',orientation="y")+
                        scale_color_brewer()+
                        scale_fill_viridis_d(option  = "cividis",labels = c("Not annotated", "Annotated"))+
                        guides(fill = guide_legend(title = "Genome type"))+

  geom_facet( data=table_info_genomelen, panel="Genome length",
                      geom = geom_point2,     #geom_point2
                      mapping = aes(x = seq_length/1000000000), color = "blue4", alpha = 0.5, size = 2,  stat='identity',
                      position= position_jitter(height=0.5,width=0) )+
  
  geom_facet( data=table_info_genomelen, panel="Number of genes",
                      geom = geom_point2,     #geom_point2
                      mapping = aes(x = gene_counts.total/1000), color = "green4",size=2, alpha = 0.5,  stat='identity',
                      position= position_jitter(height=0.5,width=0) )+
  
  
  #geom_vline( data = table_info_genomelen, aes(xintercept=1), colour="red")+
  theme_tree2(legend.position='right')+
  theme(text=element_text(family="Arial", size=25))+
  labs(x="Genome Assembly Length (Gb)")#+  xlim_expand(1, panel='Number of genomes')#+  xlim_expand(1, panel='Genome lenght')

ggtree_Arthropoda_species_3panels
```
```{r}
d <- ggimage::phylopic_uid(Arthropoda_species_Annot$order_ot)

p <- ggtree_Arthropoda_species_3panels %<+% d+
geom_tiplab(aes(image=uid), geom="phylopic", offset=2.5)

p
```

```{r}
ggsave(ggtree_Arthropoda_species_3panels, file="figs/ggtree_Arthropoda_species_3panels_colors.svg", height=18, width=16, units="in")
#sed "s/ textLength='[^']*'//" ggtree_Crustacea_species_3panels.svg > ggtree_Crustacea_species_3panels_EDIT.svg 
```

```{r}
ggsave(ggtree_Arthropoda_species_3panels, file="1.svg", height=60, width=36, units="cm")
```




```{r}
#Crustacea_genomes_firstperspp %>% dplyr::select(spp, assembly.length) %>% filter(order="Amphipoda")
how_big <- Arthropoda_genomes_firstperspp %>% dplyr::select(spp,order, seq_length)  %>% mutate(GenomeGb=seq_length/1000000000)
how_big
how_big[order(-how_big$GenomeGb),]
```
```{r}
sessionInfo()
```


<!-- ## Build tree with any species -->

<!-- ```{r} -->
<!-- resolved_names_gyllus <- rotl::tnrs_match_names(names = c("gryllus","blattodea","dipters")) -->

<!-- Gryllustree <- rotl::tol_induced_subtree(ott_ids=resolved_names_gyllus$ott_id) -->

<!-- ape::plot.phylo(Gryllustree, cex = 1.2) -->

<!-- ``` -->

<!-- ## Insects -->

<!-- ```{r} -->
<!-- Insecta_orders <- datelife::get_ott_children(ott_ids = resolved_names[resolved_names$unique_name=="Insecta",]$ott_id, ott_rank = "order") -->
<!-- Insecta_orders_subtree <- rotl::tol_induced_subtree(Insecta_orders$Insecta$ott_id) -->

<!-- Insecta_orders_subtree -->
<!-- ``` -->


<!-- ```{r, fig.height=7} -->
<!-- ape::plot.phylo(Insecta_orders_subtree, cex = 1.2) -->
<!-- ``` -->